src/mbgl/renderer/layers/render_heatmap_layer.cpp

```diff
@@ -209,4 +229,261 @@ bool RenderHeatmapLayer::queryIntersectsFeature(const GeometryCoordinates& query
     return false;
 }
 
+#if MLN_DRAWABLE_RENDERER
+void activateRenderTarget(const RenderTargetPtr& renderTarget_, bool activate, UniqueChangeRequestVec& changes) {
+    if (renderTarget_) {
+        if (activate) {
+            // The RenderTree has determined this render target should be included in the renderable set for a frame
+            changes.emplace_back(std::make_unique<AddRenderTargetRequest>(renderTarget_));
+        } else {
+            // The RenderTree is informing us we should not render anything
+            changes.emplace_back(std::make_unique<RemoveRenderTargetRequest>(renderTarget_));
+        }
+    }
+}
+
+void RenderHeatmapLayer::markLayerRenderable(bool willRender, UniqueChangeRequestVec& changes) {
+    RenderLayer::markLayerRenderable(willRender, changes);
+    activateRenderTarget(renderTarget, willRender, changes);
+}
+
+void RenderHeatmapLayer::removeTile(RenderPass renderPass, const OverscaledTileID& tileID) {
+    auto* tileLayerGroup = static_cast<TileLayerGroup*>(renderTarget->getLayerGroup(0).get());
+    stats.drawablesRemoved += tileLayerGroup->removeDrawables(renderPass, tileID).size();
+}
+
+void RenderHeatmapLayer::removeAllDrawables() {
+    RenderLayer::removeAllDrawables();
+    if (renderTarget) {
+        stats.drawablesRemoved += renderTarget->getLayerGroup(0)->getDrawableCount();
+        renderTarget->getLayerGroup(0)->clearDrawables();
+    }
+}
+
+struct alignas(16) HeatmapInterpolateUBO {
+    float weight_t;
+    float radius_t;
+    std::array<float, 2> padding;
+};
+static_assert(sizeof(HeatmapInterpolateUBO) % 16 == 0);
+
+static const std::string HeatmapShaderGroupName = "HeatmapShader";
+static const std::string HeatmapTextureShaderGroupName = "HeatmapTextureShader";
+static constexpr std::string_view HeatmapInterpolateUBOName = "HeatmapInterpolateUBO";
+
+void RenderHeatmapLayer::update(gfx::ShaderRegistry& shaders,
+                                gfx::Context& context,
+                                const TransformState& state,
+                                [[maybe_unused]] const RenderTree& renderTree,
+                                [[maybe_unused]] UniqueChangeRequestVec& changes) {
+    std::unique_lock<std::mutex> guard(mutex);
+
+    if (!renderTiles || renderTiles->empty()) {
+        removeAllDrawables();
+        return;
+    }
+
+    // Set up a render target
+    if (!renderTarget) {
+        // const auto& viewportSize = parameters.staticData.backendSize;
+        const auto size = Size(500, 500); // Size{viewportSize.width / 4, viewportSize.height / 4};
```

Comment:
I still have this know issue, not sure where to set the size of the render target.

Reply:
`state.getSize()` should be the viewport.

Reply:
It's curious to use the viewport for that.  Normally you'd pick a fixed size, like 256 or so, and just have that backed into your tile size for your source.  At least that's how we'd do it.

Reply:
Thanks @mwilsnd. `getSize()` returns the view size, which ignores the scale factor, so I divided by 2, instead of 4.